### ISSUE TITLE
Zero-mAP bug: P, R and mAP are all 0.0

### ISSUE THREAD

**Issue by athrunsunny** [ultralytics/yolov5#9059]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I train my custom one class dataset（about 40,000 pictures）use default cfg of yolov5s with the latest version（6.2）,P, R and mAP are all 0，(It's the same after about 100 epochs)however, it is normal to use the previous version（6.1）and it's normal when I train with coco128,I think this may be a bug, but I didn't debug it.

**Comment by giuseppebrb** [comment]
It's a bug already happened the other day but @glenn-jocher fixed it, today reappeared

**Comment by pourmand1376** [comment]
Have you updated your repo? @glenn-jocher has created a PR for this yesterday.

**Comment by giuseppebrb** [comment]
In my case, yesterday everything was working again, today bug returned. I clone the repository at every run in my Google Colab notebook.

**Comment by glenn-jocher** [comment]
@athrunsunny @giuseppebrb @pourmand1376 I'm still seeing this myself, but it's not consistently reproducible. I trained 10 models in a row in Colab just now and the zero-mAP bug randomly appears in 2 out of 10. Importantly **all val losses are identical**, all single-gpu trainings are now 100% reproducible, so the bug is not in training, it is in the metrics apparently. 

**I need help debugging this, I don't know what's causing the problem.**

```
# Train 10x YOLOv5s on COCO128 for 3 epochs
for _ in range(10):
  !python train.py --img 640 --batch 16 --epochs 3 --data coco128.yaml --weights yolov5s.pt --cache
```

<img width="1742" alt="Screenshot 2022-08-21 at 14 13 14" src="https://user-images.githubusercontent.com/26833433/185790367-9f9ad82e-838f-4b77-bb53-dde15e8470cb.png">

@AyushExel FYI this is happening, don't have a solution yet.

**Comment by pourmand1376** [comment]
@glenn-jocher 
Okay, If you don't know where to start, I suggest using `git bisect`. This one will find the exact commit which has made the issue. 

I will try to solve the problem. However, I don't know any commit which doesn't have this issue. Do you know any?

**Comment by glenn-jocher** [comment]
@pourmand1376 yes I've been manually bisecting commits, but since it's not 100% reproducible I would only test one training, and the most recent commit that passed the one training I would revert to. This only appeared in the last few days. Some other observations:

- --device cpu trainings always appear to work correctly
- --noplots trainings always appear to work correctly even on GPU
- another user mentioned nan batchnorm values, but this doesn't make sense to me as all val losses are identical for both zero map and normal runs.

**Comment by AyushExel** [comment]
This seems like a p0 problem. Let's prioritize this. 
I've seen similar things happen when some hyper-param is set to a bad value. Was there any commit lately related to hyps?

**Comment by glenn-jocher** [comment]
@pourmand1376 so it seems like the git bisect approach is correct, but the test must be high statistics, i.e. run 30 trainings with 30 good results and we can be 97% sure there's no bug in that commit.

**Comment by glenn-jocher** [comment]
@AyushExel no, last hyp changes were months ago. I think this might be related to EMA, running val.py directly from train.py, maybe AMP/FP16 ops. But the strangest part is that **val losses are always identical across all models, regardless of whether map is 0 or correct.**

**Comment by glenn-jocher** [comment]
Another thing I suspected was the introduction of identical filenames in /classify. This has never happened before in the repo. If that's the case then the bug would appear directly in the v6.2 tag.

**Comment by pourmand1376** [comment]
Hmm, That's correct. I am using 
```bash
for _ in range(10):
  !python train.py --img 640 --batch 16 --epochs 2 --data coco128.yaml --weights yolov5x.pt --cache
```

So far:
```bash
!git bisect good 20f1b7ea086fb317da93d2e603c4def2ebfb8187
```

**Comment by glenn-jocher** [comment]
@pourmand1376 oh, YOLOv5x will be slow. You can use the default Colab training command, the bug will appear there:
```python
for _ in range(10):
  !python train.py --img 640 --batch 16 --epochs 3 --data coco128.yaml --weights yolov5s.pt --cache
```

**Comment by glenn-jocher** [comment]
I trained v6.2 tag 15x and all trainings are correct. So it's not actually in the v6.2 release apparently but has been introduced in the 38 commits since then.

**Comment by pourmand1376** [comment]
I am updating the git bisect commits. Good ones are not for sure as I am running 4 concurrent colab sessions but bad ones are fore sure.

**Comment by robotwhispering** [comment]
I'm still on commit d1dfcab60493e80441cb00be33cfd23aabc26f6b and no issues with loss, P, R, mAP, etc.
I did notice that it depended on where I was seeing the values displayed whether they appeared as NaN or 0...and trainning didn not halt...in fact, it felt normal, other than not seeing.

**Comment by pourmand1376** [comment]
This is what bisect reports ... 

61adf017f231f470afca2636f1f13e4cce13914b is the first bad commit
```
commit 61adf017f231f470afca2636f1f13e4cce13914b
Author: Glenn Jocher <[glenn.jocher@ultralytics.com](mailto:glenn.jocher@ultralytics.com)>
Date:   Thu Aug 18 20:12:33 2022 +0200

    `torch.empty()` for speed improvements (#9025)
    
    `torch.empty()` for speed improvement
    
    Signed-off-by: Glenn Jocher <[glenn.jocher@ultralytics.com](mailto:glenn.jocher@ultralytics.com)>

:040000 040000 c7c88a2192877d503c0ef118407ca9ba9062e3f3 03d4349010436426c5870f8cab61b453890f5aeb M	models
:040000 040000 9c18cbfcbcf0c2c58e9d83d91cfdc31bd652d6b0 38a9925a4fe83db85246725490edbd5455cedb1e M	utils
```

**Comment by glenn-jocher** [comment]
@pourmand1376 ohhhhhhh this makes sense since torch.empty is not reproducible! Awesome @pourmand1376 this is a huge help. I'll revert this commit immediately, the speed gains are super small from the torch.zeros to torch.empty switch in any case.

**Comment by AyushExel** [comment]
@glenn-jocher @pourmand1376 quick question - do we know WHY this caused issues? From the [docs](https://pytorch.org/docs/stable/generated/torch.empty.html) I understand that it creates unintailized tensor with random values(explains non-reproduciblity) but I don't get why the metrics would be stuck at 0 ?

**Comment by glenn-jocher** [comment]
@pourmand1376 thanks for the help!! I've pushed #9068 with a fix by @0zppd just now

**Comment by pourmand1376** [comment]
@AyushExel 
I don't know. Maybe this one should be reported as a bug in the upstream `pytorch`!

**Comment by glenn-jocher** [comment]
@giuseppebrb @athrunsunny @pourmand1376 @robotwhispering @AyushExel good news 😃! Your original issue may now be fixed ✅ in PR #9068 by @0zppd. @pourmand1376 tracked down the problem using `git bisect` and running each training 10x since the bug was not reproducible on every training, but showed up in maybe 1/3 of the trainings. 

The original issue was that I had replaced torch.zeros() with torch.empty() on some ops like warmup and profiling to try to get some slight speed improvements, and once op in particular ran a torch.empty() tensor through the model when it was in .train() mode, leading the batchnorm layers to add those values to the tracked statistics. Since torch.empty is not initialized it can take on extremely high or low values, leading some batchnorm layers to randomly output NaN. 

The PR has been extensively tested on 10x Colab trainings and all 10 came back good now:
<img width="1211" alt="Screenshot 2022-08-21 at 15 40 00" src="https://user-images.githubusercontent.com/26833433/185793959-986ad082-c181-49b8-a564-b96270ebc409.png">

To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

**Comment by athrunsunny** [comment]
@glenn-jocher @pourmand1376  @0zppd Thank you for your great work

**Comment by shubhambagwari** [comment]
I forked the latest version of Yv5, Still, I am facing this error. 
Earlier I trained this model with 2000 images but faced the same issue. 

Roboflow has been used to label the custom dataset.
I have read the earlier discussions but not working.

![image](https://user-images.githubusercontent.com/25363312/207236797-e2cf2e4d-67ec-4a52-8fdb-ff048cf0caa3.png)

**Comment by glenn-jocher** [comment]
@shubhambagwari try 300 epochs.

**Comment by Davegdd** [comment]
> @shubhambagwari try 300 epochs.

Thank you for your answer. I was having the same issue, it resolved after increasing epochs just beyond 3.

**Comment by umarbutler** [comment]
> > @shubhambagwari try 300 epochs.
> 
> Thank you for your answer. I was having the same issue, it resolved after increasing epochs just beyond 3.

Thanks for this, I quit my training a number of times to try and fix getting 0 P, R and mAP values in the first two epochs not realising that it would only start to increase by epoch 3 😆
